python/lsst/ip/isr/linearize.py

```diff
@@ -0,0 +1,228 @@
+#
+# LSST Data Management System
+# Copyright 2016 AURA/LSST.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+import abc
+from itertools import izip
+
+import numpy as np
+
+from lsst.pipe.base import Struct
+from .isrLib import applyLookupTable
+
+__all__ = ["LinearizeBase", "LinearizeLookupTable", "LinearizeSquared"]
+
+class LinearizeBase(object):
+    """Abstract base class functor for correcting non-linearity
+    """
+    __metaclass__ = abc.ABCMeta
+    LinearityType = None
+
+    @abc.abstractmethod
+    def __call__(self, image, detector, log=None):
+        """Correct non-linearity
+
+        @param[in] image  image to be corrected (an lsst.afw.image.Image)
+        @param[in] detector  detector information (an instance of lsst::afw::cameraGeom::Detector)
+        @param[in] log  logger (an lsst.pex.logging.Log), or None to disable logging;
+                    a warning is logged if amplifiers are skipped or other worrisome events occur
+
+        @return an lsst.pipe.base.Struct containing at least the following fields:
+        - numAmps number of amplifiers found
+        - numLinearized  number of amplifiers linearized
```

Comment:
You raise a good point, but I respectfully disagree with your conclusion. The heart of the problem is that there ought to be only one linearity type for the whole detector, but I know of no good way to represent table-wide properties in a catalog. A bad way is to use the table's metadata, but that is not type safe. Another alternative is to make linearity type a property of `Detector` instead of `AmpInfoCatalog`, but that separates related data. Of these three options, I prefer the one I chose.

Getting back to the linearizer...I want to persist a whole detector's worth of lookup table for each linearizer. This feels logical to me, since the whole detector will be corrected in one pass through the ISR task, and strikes a good balance between persisting too much information in one blob (the whole focal plane worth of lookup tables) and too little (one amplifier's lookup table). Having done this, as a functor, I think it is logical for `__call__` to correct the whole CCD image. Finally, `Detector` contains some information that is useful, and it would be annoying to have to pass in both a detector and an amp info catalog.